# 4.8 TASK

# =================================================

# In this TASK, find code and output for Steps 1-7 as requested.

# =================================================

# 1. Create a new notebook for this task. Be sure to import the relevant libraries, along with your ords_prods_merge dataframe.

In [1]:
# Importing libraries

import pandas as pd
import numpy as np
import os

In [2]:
# Create path.

path = r'/Users/matthewschleigh/Desktop/CareerFoundry/Achievement 4/Instacart Basket Analysis'

In [4]:
# Import ords_prods_merge

ords_prods_merge = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'ords_prods_merge_4_8_ex'))

In [5]:
# Check output.

ords_prods_merge.head()

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0   2539329        1             1                  2                  8   
1   2398795        1             2                  3                  7   
2    473747        1             3                  3                 12   
3   2254736        1             4                  4                  7   
4    431534        1             5                  4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered _merge  \
0                     0.0         196                  1          0   both   
1                    15.0         196                  1          1   both   
2                    21.0         196                  1          1   both   
3                    29.0         196                  1          1   both   
4                    28.0         196                  1          1   both   

  product_name  aisle_id  department_id  prices    price_range_loc  \
0         Soda        77              7     9.0  Mid-range product   
1         Soda        77              7     9.0  Mid-range product   
2         Soda        77              7     9.0  Mid-range product   
3         Soda        77              7     9.0  Mid-range product   
4         Soda        77              7     9.0  Mid-range product   

      busiest_day    busiest_days busiest_period_of_day  max_order  \
0  Regularly busy  Regularly busy        Average orders         10   
1  Regularly busy     Slowest day        Average orders         10   
2  Regularly busy     Slowest day        Average orders         10   
3      Least busy     Slowest day        Average orders         10   
4      Least busy     Slowest day           Most orders         10   

   loyalty_flag  
0  New customer  
1  New customer  
2  New customer  
3  New customer  
4  New customer

In [6]:
ords_prods_merge.shape

(32404859, 20)

# 2. In this Exercise, you learned how to find the aggregated mean of the “order_number” column grouped by “department_id” for a subset of your dataframe. Now, repeat this process for the entire dataframe.

In [7]:
# Use agg() function to determine mean for entire df.

ords_prods_merge.groupby('department_id').agg({'order_number': ['mean']})

order_number
                      mean
department_id             
1                15.457838
2                17.277920
3                17.170395
4                17.811403
5                15.215751
6                16.439806
7                17.225802
8                15.340650
9                15.895474
10               20.197148
11               16.170638
12               15.887671
13               16.583536
14               16.773669
15               16.165037
16               17.665606
17               15.694469
18               19.310397
19               17.177343
20               16.473447
21               22.902379

# 3. Analyze the result. How do the results for the entire dataframe differ from those of the subset? Include your comments in a markdown cell below the executed code.

# Since this is the entire df and not simply a sampling of 1 million lines, we see a row for each of the departments included in our dataframe. Looking closer, the mean values determined by the aggregate function differ from those in the sampling in the exercise, because the program is calculating based off of the entire dataframe rather than just a sample. We should expect these mean values to be different, otherwise we should determine why they're the same. The largest difference between means is found within deparment 17, with 11.29 in the sample and 15.69 across the entire df.

# 4. Follow the instructions in the Exercise for creating a loyalty flag for existing customers using the transform() and loc() functions.

In [9]:
# Loyalty flag with 3 types previously created in exercise, then saved when exported.
# Check output.

ords_prods_merge['loyalty_flag'].value_counts(dropna=False)

Regular customer    15876776
Loyal customer      10284093
New customer         6243990
Name: loyalty_flag, dtype: int64

In [10]:
# Scroll to far right side to see new loyalty_flag column.

ords_prods_merge.head()

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0   2539329        1             1                  2                  8   
1   2398795        1             2                  3                  7   
2    473747        1             3                  3                 12   
3   2254736        1             4                  4                  7   
4    431534        1             5                  4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered _merge  \
0                     0.0         196                  1          0   both   
1                    15.0         196                  1          1   both   
2                    21.0         196                  1          1   both   
3                    29.0         196                  1          1   both   
4                    28.0         196                  1          1   both   

  product_name  aisle_id  department_id  prices    price_range_loc  \
0         Soda        77              7     9.0  Mid-range product   
1         Soda        77              7     9.0  Mid-range product   
2         Soda        77              7     9.0  Mid-range product   
3         Soda        77              7     9.0  Mid-range product   
4         Soda        77              7     9.0  Mid-range product   

      busiest_day    busiest_days busiest_period_of_day  max_order  \
0  Regularly busy  Regularly busy        Average orders         10   
1  Regularly busy     Slowest day        Average orders         10   
2  Regularly busy     Slowest day        Average orders         10   
3      Least busy     Slowest day        Average orders         10   
4      Least busy     Slowest day           Most orders         10   

   loyalty_flag  
0  New customer  
1  New customer  
2  New customer  
3  New customer  
4  New customer

# 5. Use the loyalty flag you created and check the basic statistics of the product prices for each loyalty category. What you’re trying to determine is whether the prices of products purchased by loyal customers differ from those purchased by regular or new customers.

In [14]:
# Use agg() function to find mean, min, and max values for the 'prices' column related to 'loyalty_flag' categories.

ords_prods_merge.groupby('loyalty_flag').agg({'prices': ['mean', 'max', 'min']})

prices              
                       mean      max  min
loyalty_flag                             
Loyal customer    10.386336  99999.0  1.0
New customer      13.294670  99999.0  1.0
Regular customer  12.495717  99999.0  1.0

# 6. Create a spending flag for each user based on the average price across all their orders.

In [15]:
# Determine mean spending amounts for each customer.

ords_prods_merge['mean_spending'] = ords_prods_merge.groupby(['user_id'])['prices'].transform(np.mean)

In [16]:
# Check output.

ords_prods_merge.head()

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0   2539329        1             1                  2                  8   
1   2398795        1             2                  3                  7   
2    473747        1             3                  3                 12   
3   2254736        1             4                  4                  7   
4    431534        1             5                  4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered _merge  \
0                     0.0         196                  1          0   both   
1                    15.0         196                  1          1   both   
2                    21.0         196                  1          1   both   
3                    29.0         196                  1          1   both   
4                    28.0         196                  1          1   both   

   ... aisle_id  department_id  prices    price_range_loc     busiest_day  \
0  ...       77              7     9.0  Mid-range product  Regularly busy   
1  ...       77              7     9.0  Mid-range product  Regularly busy   
2  ...       77              7     9.0  Mid-range product  Regularly busy   
3  ...       77              7     9.0  Mid-range product      Least busy   
4  ...       77              7     9.0  Mid-range product      Least busy   

     busiest_days busiest_period_of_day max_order  loyalty_flag mean_spending  
0  Regularly busy        Average orders        10  New customer      6.367797  
1     Slowest day        Average orders        10  New customer      6.367797  
2     Slowest day        Average orders        10  New customer      6.367797  
3     Slowest day        Average orders        10  New customer      6.367797  
4     Slowest day           Most orders        10  New customer      6.367797  

[5 rows x 21 columns]

In [17]:
# Create flag with 2 types of criteria: <10 = 'Low_spender', >=10 = 'High_spender'

ords_prods_merge.loc[ords_prods_merge['mean_spending'] < 10, 'Spender_type'] = 'Low_spender'

In [18]:
ords_prods_merge.loc[ords_prods_merge['mean_spending'] >= 10, 'Spender_type'] = 'High_spender'

In [19]:
# Check output.

ords_prods_merge['Spender_type'].value_counts(dropna=False)

Low_spender     31770614
High_spender      634245
Name: Spender_type, dtype: int64

In [20]:
ords_prods_merge.head()

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0   2539329        1             1                  2                  8   
1   2398795        1             2                  3                  7   
2    473747        1             3                  3                 12   
3   2254736        1             4                  4                  7   
4    431534        1             5                  4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered _merge  \
0                     0.0         196                  1          0   both   
1                    15.0         196                  1          1   both   
2                    21.0         196                  1          1   both   
3                    29.0         196                  1          1   both   
4                    28.0         196                  1          1   both   

   ... department_id  prices    price_range_loc     busiest_day  \
0  ...             7     9.0  Mid-range product  Regularly busy   
1  ...             7     9.0  Mid-range product  Regularly busy   
2  ...             7     9.0  Mid-range product  Regularly busy   
3  ...             7     9.0  Mid-range product      Least busy   
4  ...             7     9.0  Mid-range product      Least busy   

     busiest_days busiest_period_of_day max_order  loyalty_flag  \
0  Regularly busy        Average orders        10  New customer   
1     Slowest day        Average orders        10  New customer   
2     Slowest day        Average orders        10  New customer   
3     Slowest day        Average orders        10  New customer   
4     Slowest day           Most orders        10  New customer   

   mean_spending Spender_type  
0       6.367797  Low_spender  
1       6.367797  Low_spender  
2       6.367797  Low_spender  
3       6.367797  Low_spender  
4       6.367797  Low_spender  

[5 rows x 22 columns]

# 7. Create an order frequency flag that marks the regularity of a user’s ordering behavior according to the median in the “days_since_prior_order” column. 

In [21]:
# Determine median values for 'days_since_prior_order' column.

ords_prods_merge['median_ord_freq'] = ords_prods_merge.groupby(['user_id'])['days_since_prior_order'].transform(np.median)

In [22]:
# Check output.

ords_prods_merge.head()

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0   2539329        1             1                  2                  8   
1   2398795        1             2                  3                  7   
2    473747        1             3                  3                 12   
3   2254736        1             4                  4                  7   
4    431534        1             5                  4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered _merge  \
0                     0.0         196                  1          0   both   
1                    15.0         196                  1          1   both   
2                    21.0         196                  1          1   both   
3                    29.0         196                  1          1   both   
4                    28.0         196                  1          1   both   

   ... prices    price_range_loc     busiest_day    busiest_days  \
0  ...    9.0  Mid-range product  Regularly busy  Regularly busy   
1  ...    9.0  Mid-range product  Regularly busy     Slowest day   
2  ...    9.0  Mid-range product  Regularly busy     Slowest day   
3  ...    9.0  Mid-range product      Least busy     Slowest day   
4  ...    9.0  Mid-range product      Least busy     Slowest day   

  busiest_period_of_day max_order  loyalty_flag mean_spending  Spender_type  \
0        Average orders        10  New customer      6.367797   Low_spender   
1        Average orders        10  New customer      6.367797   Low_spender   
2        Average orders        10  New customer      6.367797   Low_spender   
3        Average orders        10  New customer      6.367797   Low_spender   
4           Most orders        10  New customer      6.367797   Low_spender   

  median_ord_freq  
0            20.0  
1            20.0  
2            20.0  
3            20.0  
4            20.0  

[5 rows x 23 columns]

In [23]:
# Create flag for reorder frequency with 3 types:
#   >20         = 'Non_frequent customer'
#   >10 & <= 20 = 'Regular customer'
#   <=10        = 'Frequent customer'

ords_prods_merge.loc[ords_prods_merge['median_ord_freq'] > 20, 'customer_frequency'] = 'Non-frequent customer'

In [24]:
ords_prods_merge.loc[(ords_prods_merge['median_ord_freq'] > 10) & (ords_prods_merge['median_ord_freq'] <= 20), 'customer_frequency'] = 'Regular customer'

In [25]:
ords_prods_merge.loc[ords_prods_merge['median_ord_freq'] <= 10, 'customer_frequency'] = 'Frequent customer'

In [26]:
# Check output.

ords_prods_merge['customer_frequency'].value_counts(dropna=False)

Frequent customer        22796659
Regular customer          6921472
Non-frequent customer     2686728
Name: customer_frequency, dtype: int64

In [27]:
ords_prods_merge.head(10)

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0   2539329        1             1                  2                  8   
1   2398795        1             2                  3                  7   
2    473747        1             3                  3                 12   
3   2254736        1             4                  4                  7   
4    431534        1             5                  4                 15   
5   3367565        1             6                  2                  7   
6    550135        1             7                  1                  9   
7   3108588        1             8                  1                 14   
8   2295261        1             9                  1                 16   
9   2550362        1            10                  4                  8   

   days_since_prior_order  product_id  add_to_cart_order  reordered _merge  \
0                     0.0         196                  1          0   both   
1                    15.0         196                  1          1   both   
2                    21.0         196                  1          1   both   
3                    29.0         196                  1          1   both   
4                    28.0         196                  1          1   both   
5                    19.0         196                  1          1   both   
6                    20.0         196                  1          1   both   
7                    14.0         196                  2          1   both   
8                     0.0         196                  4          1   both   
9                    30.0         196                  1          1   both   

   ...    price_range_loc     busiest_day    busiest_days  \
0  ...  Mid-range product  Regularly busy  Regularly busy   
1  ...  Mid-range product  Regularly busy     Slowest day   
2  ...  Mid-range product  Regularly busy     Slowest day   
3  ...  Mid-range product      Least busy     Slowest day   
4  ...  Mid-range product      Least busy     Slowest day   
5  ...  Mid-range product  Regularly busy  Regularly busy   
6  ...  Mid-range product  Regularly busy     Busiest day   
7  ...  Mid-range product  Regularly busy     Busiest day   
8  ...  Mid-range product  Regularly busy     Busiest day   
9  ...  Mid-range product      Least busy     Slowest day   

   busiest_period_of_day max_order  loyalty_flag mean_spending Spender_type  \
0         Average orders        10  New customer      6.367797  Low_spender   
1         Average orders        10  New customer      6.367797  Low_spender   
2         Average orders        10  New customer      6.367797  Low_spender   
3         Average orders        10  New customer      6.367797  Low_spender   
4            Most orders        10  New customer      6.367797  Low_spender   
5         Average orders        10  New customer      6.367797  Low_spender   
6         Average orders        10  New customer      6.367797  Low_spender   
7            Most orders        10  New customer      6.367797  Low_spender   
8         Average orders        10  New customer      6.367797  Low_spender   
9         Average orders        10  New customer      6.367797  Low_spender   

   median_ord_freq customer_frequency  
0             20.0   Regular customer  
1             20.0   Regular customer  
2             20.0   Regular customer  
3             20.0   Regular customer  
4             20.0   Regular customer  
5             20.0   Regular customer  
6             20.0   Regular customer  
7             20.0   Regular customer  
8             20.0   Regular customer  
9             20.0   Regular customer  

[10 rows x 24 columns]

# Export your dataframe as a pickle file and store it correctly in your “Prepared Data” folder.

In [28]:
ords_prods_merge.to_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'ords_prods_merge_4_8_task.pkl'))